api/src/test/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesTagHandlerTest.java

```diff
@@ -75,6 +85,46 @@ public void getSubstitution_shouldAddEncounterDiagnosisElementToFormSubmissionCo
 
     }
 
+    @Test
+    public void getSubstitution_shouldAddHiddenFieldCarryingConceptSourceName() throws Exception {
+        // setup
+        when(formEntryContext.getMode()).thenReturn(FormEntryContext.Mode.ENTER);
+        when(uiUtils.includeFragment(eq("coreapps"), eq("diagnosis/encounterDiagnoses"), anyMap())).thenReturn("Some Fragment");
+        Map<String,String> attributes = new HashMap<String, String>();
+        attributes.put("required", "true");
+        attributes.put(CoreAppsConstants.HTMLFORMENTRY_ENCOUNTER_DIAGNOSES_TAG_INCLUDE_PRIOR_DIAGNOSES_ATTRIBUTE_NAME, "admit");
+        attributes.put("selectedDiagnosesTarget", "example-target");
+        attributes.put("conceptSource", "ICPC2");
+        String hiddenInputElement = "<input type=\"hidden\" id=\"concept-source\" value=\"ICPC2\"/>";
+
+        // replay
+        String generatedHtml = encounterDiagnosesTagHandler.getSubstitution(formEntrySession, formSubmissionController, attributes);
+    
+        // verify
+        assertTrue(StringUtils.contains(generatedHtml, hiddenInputElement));
+
+    }
+
+    @Test
+    public void getSubstitution_shouldNotAddHiddenConceptSourceNameFieldGivenEmptyOrNullValue() throws Exception {
+        // setup
+        when(formEntryContext.getMode()).thenReturn(FormEntryContext.Mode.ENTER);
+        when(uiUtils.includeFragment(eq("coreapps"), eq("diagnosis/encounterDiagnoses"), anyMap())).thenReturn("Some Fragment");
+        Map<String,String> attributes = new HashMap<String, String>();
+        attributes.put("required", "true");
+        attributes.put(CoreAppsConstants.HTMLFORMENTRY_ENCOUNTER_DIAGNOSES_TAG_INCLUDE_PRIOR_DIAGNOSES_ATTRIBUTE_NAME, "admit");
+        attributes.put("selectedDiagnosesTarget", "example-target");
+        attributes.put("conceptSource", "");
+        String hiddenInputElement = "\n<input type=\"hidden\" id=\"concept-source\" value=\"ICPC2\"/>\n";
+
+        // replay
+        String generatedHtml = encounterDiagnosesTagHandler.getSubstitution(formEntrySession, formSubmissionController, attributes);
+        
+        // verify
+        assertFalse(StringUtils.contains(generatedHtml, hiddenInputElement));
```

Comment:
If `generatedHtml` were to contain
```html
<input type="hidden" id="concept-source" value="ICD-10-WHO"/>
```
then this test would be misleading.

Maybe this would be "enough":
```java
String hiddenInputElement = "\n<input type=\"hidden\" id=\"concept-source\"";
```
In that case it should be renamed `hiddenInputElementStartingPart` or something like that.

A proper test would actually be to parse `generatedHtml` and ensure that there is no tag ID's `"concept-source", probably overkill.